framework/src/executioners/Transient.C

```diff
@@ -257,31 +257,42 @@ Transient::init()
   }
 
   _problem.initialSetup();
+
   _time_stepper->init();
 
   if (_app.isRestarting())
     _time_old = _time;
 
   _problem.outputStep(EXEC_INITIAL);
 
-  // If this is the first step
-  if (_t_step == 0)
-    _t_step = 1;
+  if (_app.isRecovering()) // Recover case
+  {
+    if (_t_step == 0)
+      mooseError("Internal error in Transient executioner: _t_step is equal to 0 while recovering "
+                 "in init().");
 
-  if (_t_step > 1) // Recover case
     _dt_old = _dt;
-
+  }
   else
   {
+    if (_t_step != 0)
+      mooseError("Internal error in Transient executioner: _t_step must be 0 without "
+                 "recovering in init().");
+
     computeDT();
-    //  _dt = computeConstrainedDT();
     _dt = getDT();
-  }
-  if (_dt == 0)
-    mooseError("Initial time step size is not evaluated properly");
+    if (_dt == 0)
+      mooseError("Initial time step size is not evaluated properly by the time stepper in use. A "
+                 "nonzero time step size is expected.\n"
```

Comment:
The problem with saying "time step size is not evaluated properly" is that you have to use the other sentence to tell people what is proper. Just tell them directly what happened
```
Time stepper computed time step size of zero which is not allowed.
```